compiler/qsc_qasm3/src/semantic/lowerer.rs

```diff
@@ -792,29 +814,25 @@ impl Lowerer {
         let stmt_span = stmt.span;
         let name = stmt.identifier.name.clone();
         let qsharp_ty = self.convert_semantic_type_to_qsharp_type(&ty.clone(), ty_span);
-        let symbol = Symbol {
-            name: name.to_string(),
-            ty: ty.clone(),
+        let symbol = Symbol::new(
+            &name,
+            stmt.identifier.span,
+            ty.clone(),
             qsharp_ty,
-            span: stmt.identifier.span,
-            io_kind: IOKind::Default,
-        };
+            IOKind::Default,
+        );
 
         // process the symbol and init_expr gathering any errors
         let init_expr = match init_expr {
             Some(expr) => match expr {
-                syntax::ValueExpression::Expr(expr) => semantic::ValueExpression::Expr(
-                    self.lower_expr_with_target_type(Some(expr), &ty, stmt_span),
-                ),
-                syntax::ValueExpression::Measurement(measure_expr) => {
-                    semantic::ValueExpression::Measurement(
-                        self.lower_measure_expr(measure_expr, stmt_span),
-                    )
+                syntax::ValueExpr::Expr(expr) => {
+                    self.lower_expr_with_target_type(Some(expr), &ty, stmt_span)
+                }
+                syntax::ValueExpr::Measurement(measure_expr) => {
+                    self.lower_measure_expr(measure_expr)
```

Comment:
I think we need a call to cast to target type here as well. If the types already match we just get the expr back, otherwise we get the implicit cast.